pkg/client/connection.go

```diff
@@ -96,4 +96,10 @@ type Connection interface {
 
 	// Unsubscribe unsubscribes from a destination
 	Unsubscribe(destination string) error
+
+	// Requestor API
+	NewRequestor(spec RequestorSpec) Requestor
+
+	// Responder API
+	NewResponder(spec ResponderSpec) (r Responder, err error)
```

Comment:
This is a bit asymmetric.
Currently, my Requestor subscribes and unsubscribes on each send.
I could subscribe upon creation which would make me add `error` to the API, and also add a `Close()` function that unsubscribes.
This is more symmetric, and also allows catching configuration errors early.

What do you think?

Reply:
Yes, I think that  `NewRequestor` should also return an error. That error may be just that there is a missing mandatory parameter in the spec, or an error caused when trying to start receiving messages.

I'd try to avoid using the terms subscribe/publish here, because this isn't using topics and what we usually understand when we say subscribe/publish. This is rather sending/receiving messages to/from queues.

I think that when you start the requestor it should inmediately start receiving messages from the responses queue. It should also create a map of pending requests, indexed by the request id. When it receives a response it should check in that map, to see if there is a pending request. If there is such request it should then call the callback function (maybe in a separate goroutine?). If there is no such request then it should loudly complain (in the log).

Reply:
Holding a map of the sent requests was the original implementation. 
I just removed it since we decided the callback returns the response message so I don't need to keep track of requests.
If we decide we are initializing the Requestor and subscribe immediately, I will return that piece of code.

Reply:
The response callback (or response handler) returns the response message, but this happens in the responder side. How will the requestor side correlate a message that it receives in a queue with the handler function that it needs to execute to handle it? You will need some kind of table that tells you that:

```
request_id:handler_function
123:a_function
456:another_fucnction
...
```

That can be a map or any other mechanism you consider convenient, but needs to exist.